src/energyplus/ForwardTranslator/ForwardTranslateThermalZone.cpp

```diff
@@ -169,159 +169,775 @@ namespace openstudio {
 
 namespace energyplus {
 
-  boost::optional<IdfObject> ForwardTranslator::translateThermalZone(ThermalZone& modelObject) {
-    std::string s;
+  void ForwardTranslator::translateThermalZoneSpacesWhenCombinedSpaces(model::ThermalZone& modelObject, IdfObject& idfObject) {
+
+    std::string tzName = modelObject.nameString();
+
+    std::vector<Space> spaces = modelObject.spaces();
+
+    // EXCLUDE SPACE TRANSLATION (old behavior)
+    // Note, when m_excludeSpaceTranslation = true, when you reach this point of the forward translator thermalZone.combineSpaces()
+    // has already been called. This happens in ForwardTranslator::translateModelPrivate. As a result, each zone has 0 or 1 space only
+    if (spaces.empty()) {
+      LOG(Warn, "ThermalZone " << tzName << " does not have any geometry or loads associated with it.");
+    } else {
+      OS_ASSERT(spaces.size() == 1);
+      auto& space = spaces[0];
+
+      idfObject.setDouble(openstudio::ZoneFields::DirectionofRelativeNorth, spaces[0].directionofRelativeNorth());
+      idfObject.setDouble(openstudio::ZoneFields::XOrigin, space.xOrigin());
+      idfObject.setDouble(openstudio::ZoneFields::YOrigin, space.yOrigin());
+      idfObject.setDouble(openstudio::ZoneFields::ZOrigin, space.zOrigin());
+
+      if (space.partofTotalFloorArea()) {
+        idfObject.setString(openstudio::ZoneFields::PartofTotalFloorArea, "Yes");
+      } else {
+        idfObject.setString(openstudio::ZoneFields::PartofTotalFloorArea, "No");
+      }
+
+      // translate the space now
+      translateAndMapModelObject(space);
+
+      // Translate all Space-specific loads
+      auto translateSpaceLoads = [this](auto loads) {
+        std::sort(loads.begin(), loads.end(), WorkspaceObjectNameLess());
+        for (auto& load : loads) {
+          translateAndMapModelObject(load);
+        }
+      };
+
+      // TODO: THIS SHOULD HAPPEN. The translateSpace should do it!
+      translateSpaceLoads(space.shadingSurfaceGroups());
```

Comment:
Why can't we just let translateSpace to this work for both cases m_excludeSpaceTranslation == true and m_excludeSpaceTranslation == false